### PR TITLE
feature/MSSDK1-12

### PR DESCRIPTION
**Pull request by duprijil**
Add loading MP from GitHub

### CODE REVIEW

**Review by schivmeister — CHANGES_REQUESTED** [review]
Some observations made. Tested some corner cases but was not able to test an actual repo load:

```sh
mapping_packages = load_mapping_packages_from_github(github_repository_url="https://github.com/OP-TED/ted-rdf-mapping-eforms",packages_path_pattern="mappings/*_can_*",branch_or_tag_name=None) # or "develop" which is this repo's default branch
```

Please see if that works for you and let me know what I missed if it does. Otherwise, feel free to ignore all suggestions or to slate them for later incorporation, as none of them are critical now (other than the above working).

File: tests/unit/adapters/test_github_extractor.py [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+            pass
+
+
+def test_github_extract_fails_on_not_existing_repo_url(dummy_github_branch_name: str,
+                                                       dummy_repo_package_path: Path,
+                                                       dummy_packages_path_pattern: str) -> None:
+    with tempfile.TemporaryDirectory() as tmp_dir:
+        tmp_dir_path = Path(tmp_dir)
+        with pytest.raises(ValueError):
+            with GithubPackageExtractor().extract(repository_url="",
```
[the remark this answers is not in the document]
Comment (duprijil) [reply tests/unit/adapters/test_github_extractor.py:119]: done

File: tests/unit/adapters/test_github_extractor.py [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+from mapping_suite_sdk.adapters.extractor import GithubPackageExtractor
+
+
+def test_github_extractor_success_with_default_args(dummy_github_project_path: Path,
+                                                    dummy_github_branch_name: str,
+                                                    dummy_repo_package_path: Path) -> None:
+    # Setup
+    with tempfile.TemporaryDirectory() as tmp_dir:
+        tmp_dir_path = Path(tmp_dir)
+        repo_path = tmp_dir_path / dummy_github_project_path.name
```
[the remark this answers is not in the document]
Comment (duprijil) [reply tests/unit/adapters/test_github_extractor.py:17]: If I add online repo, this will test the ability of the library (GitPython) to understand local or online repo, and this will not test the logic developed in sdk itself. So let's skip that.

File: mapping_suite_sdk/services/load_mapping_package.py [hunk 3]
```diff
@@ -77,9 +77,118 @@ def load_mapping_package_from_archive(
...
 
     with archive_unpacker.extract_temporary(mapping_package_archive_path) as temp_mapping_package_folder_path:
 
         return load_mapping_package_from_folder(mapping_package_folder_path=temp_mapping_package_folder_path,
                                                 mapping_package_loader=mapping_package_loader)
+
+
+def load_mapping_packages_from_github(
+        github_repository_url: str,
+        packages_path_pattern: str,
```
[the remark this answers is not in the document]
Comment (duprijil) [reply mapping_suite_sdk/services/load_mapping_package.py:90]: I would leave this mandatory. At least I can do another service/services that takes the packages from OP eForms or standard forms repos. In this case, you can add only branch or tag. Even tag can be 'develop' as default.

[the remark this answers is not in the document]
Comment (duprijil) [reply mapping_suite_sdk/services/load_mapping_package.py:90]: But this is a specific need, to get packages from OP with current structure.

File: mapping_suite_sdk/adapters/extractor.py [hunk 4]
```diff
@@ -139,3 +196,143 @@ def pack_directory(source_dir: Path, output_path: Path) -> Path:
...
+        """
+
+        with tempfile.TemporaryDirectory() as temp_dir:
+            temp_dir_path = Path(temp_dir)
+            try:
+                Repo.clone_from(repository_url, temp_dir_path, branch=branch_or_tag_name, depth=1)
+                yield [package_path for package_path in temp_dir_path.glob(packages_path_pattern) if
+                       package_path.is_dir()]
+            except Exception as e:
+                raise ValueError(f"Failed to clone repository: {e}")
```
[the remark this answers is not in the document]
Comment (duprijil) [reply mapping_suite_sdk/adapters/extractor.py:338]: done

File: mapping_suite_sdk/services/load_mapping_package.py [hunk 5]
```diff
@@ -77,9 +77,118 @@ def load_mapping_package_from_archive(
...
+        branch_or_tag_name: str,
```
[the remark this answers is not in the document]
Comment (duprijil) [reply mapping_suite_sdk/services/load_mapping_package.py:91]: Agree. I made this mandatory because of the specific of library. If you pass `branch=None` to `Repo.clone_from()`, it will not work. But it's bad that I decided to do the logic depending on the library. I will change that, so `branch_or_tag_name` will be optional.

File: mapping_suite_sdk/services/load_mapping_package.py [hunk 6]
```diff
@@ -77,9 +77,118 @@ def load_mapping_package_from_archive(
...
+    3. Loads each matching directory as a mapping package
+    4. Cleans up temporary files
+    5. Returns the list of loaded packages
+
+    Args:
+        github_repository_url: The URL of the GitHub repository. Must be a valid GitHub
+            repository URL (e.g., "https://github.com/org/repo").
+        packages_path_pattern: Glob pattern to match package paths within the repository.
+            The pattern is relative to the repository root and supports glob-style
+            matching (e.g., "mappings/package*" or "mappings/*_can_*").
```
[the remark this answers is not in the document]
Comment (duprijil) [reply mapping_suite_sdk/services/load_mapping_package.py:114]: Yes. I've added a test for that check.

File: tests/unit/adapters/test_github_extractor.py [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+def test_github_extract_temporary_fails_on_bad_pattern(dummy_github_project_path: Path,
+                                                       dummy_github_branch_name: str) -> None:
+    with pytest.raises(ValueError):
+        with GithubPackageExtractor().extract_temporary(repository_url=dummy_github_project_path,
+                                                        packages_path_pattern="wrong*_pattern",
+                                                        branch_or_tag_name=dummy_github_branch_name):
+            pass
+
+
+def test_github_extract_temporary_fails_on_wrong_tag_name(dummy_github_project_path: Path,
```
[the remark this answers is not in the document]
Comment (duprijil) [reply tests/unit/adapters/test_github_extractor.py:76]: The pattern should be mandatory, or we have a specific service that has a default structure.
The `branch_or_tag_name=None` is solved in previous comment.
https://github.com/meaningfy-ws/mapping-suite-sdk/pull/8#discussion_r1971972366

File: mapping_suite_sdk/services/load_mapping_package.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (schivmeister) [reply mapping_suite_sdk/services/load_mapping_package.py:90]: Let's leave it mandatory then, I realize we shouldn't be assuming any specific pattern, you are right. Let's not also have any default for branch or tag, let it be nullable in which case the Git client will resolve the default branch.

**Review by schivmeister — APPROVED** [review]
Thanks for attending to the feedback. Some non-critical responses given FYI, but otherwise good to go.

File: mapping_suite_sdk/adapters/extractor.py [hunk 8]
```diff
@@ -241,12 +241,12 @@ def extract(
     ) -> Path:
         """Extract a specific package from a GitHub repository.
 
-        This method clones a GitHub repository to a specified destination and returns
+        This method clones a public GitHub repository to a specified destination and returns
```
[the remark this answers is not in the document]
Comment (duprijil) [reply mapping_suite_sdk/adapters/extractor.py:244]: I will delete public

File: tests/unit/adapters/test_github_extractor.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (duprijil) [reply tests/unit/adapters/test_github_extractor.py:74]: added test
